source/communication_interfaces/src/sockets/UDPSocket.cpp

```diff
@@ -0,0 +1,135 @@
+  this->server_fd_ = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+  if (this->server_fd_ < 0) {
+    throw exceptions::SocketConfigurationException("Opening socket failed");
+  }
+  if (this->enable_reuse_ && this->enable_reuse_->get_value()) {
+    const int opt_reuse = 1;
+    if (setsockopt(this->server_fd_, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt_reuse, sizeof(opt_reuse)) != 0) {
+      throw exceptions::SocketConfigurationException("Setting socket option (enable reuse) failed");
+    }
+  }
+  if (bind_socket) {
+    if (bind(this->server_fd_, (sockaddr*) &(this->server_address_), sizeof(server_address_)) != 0) {
+      throw exceptions::SocketConfigurationException("Binding socket failed.");
+    }
+  }
+
+  if (this->timeout_duration_sec_->get_value() > 0.0) {
+    timeval timeout{};
+    auto secs = std::floor(this->timeout_duration_sec_->get_value());
+    timeout.tv_sec = static_cast<long>(secs);
+    timeout.tv_usec = static_cast<long>((this->timeout_duration_sec_->get_value() - secs) * 1e6);
+    if (setsockopt(this->server_fd_, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout)) != 0) {
+      throw exceptions::SocketConfigurationException("Setting socket timeout failed.");
+    }
+  }
+}
+
+bool UDPSocket::recvfrom(const sockaddr_in& address, ByteArray& buffer) {
+  std::vector<char> local_buffer(this->buffer_size_->get_value());
+  auto receive_length = ::recvfrom(
+      this->server_fd_, local_buffer.data(), this->buffer_size_->get_value(), 0, (sockaddr*) &(address),
+      &(this->addr_len_));
+  if (receive_length < 0) {
+    return false;
+  }
+  local_buffer.at(receive_length) = 0;
+  buffer.reset();
+  buffer.load(local_buffer.data(), receive_length);
+  return true;
+}
+
+bool UDPSocket::sendto(const sockaddr_in& address, const ByteArray& buffer) const {
+  std::vector<char> local_buffer;
+  buffer.copy_to(local_buffer);
+  int send_length = ::sendto(
+      this->server_fd_, local_buffer.data(), local_buffer.size(), 0, (sockaddr*) &(address), this->addr_len_);
+  return send_length == static_cast<int>(local_buffer.size());
+}
+
+void UDPSocket::validate_and_set_parameter(const std::shared_ptr<ParameterInterface>& parameter) {
+  // TODO can we put that in ParameterMap?
+  if (this->parameters_.find(parameter->get_name()) == this->parameters_.end()) {
+    throw state_representation::exceptions::InvalidParameterException(
+        "Invalid parameter '" + parameter->get_name() + "' for class UDPSocket.");
+  }
+  this->assert_parameter_valid(parameter);
+  if (parameter->is_empty()) {
+    throw state_representation::exceptions::InvalidParameterException(
+        "Parameter '" + parameter->get_name() + "' cannot be empty.");
+  }
+  if (parameter->get_name() == "timeout_duration_sec" && parameter->get_parameter_value<double>() < 0.0) {
+    throw state_representation::exceptions::InvalidParameterException(
+        "Parameter 'timeout_duration_sec' cannot be negative.");
+  }
+  // TODO can we put that in ParameterMap?
```

Comment:
Same here, it's a bit annoying to do that switch case each time but here I'm not sure it would be sensible to implement a helper for it.

Reply:
hmm, it's a tricky question. Generally all member functions that update the `parameters_` property should result in `validate_and_set_parameter` being called, and it's only at the end of validation that the parameter is actually set. It should not be encouraged to modify the parameter map through other means because then you risk skipping the validations. For that reason, defining an `update_parameter_value` / `copy_parameter_value` helper with a switch case is tricky because you might be tempted to call it in other places, or else be required to add more validation to that helper again.

I think originally the design intent was as follows:
```c++
// set a parameter at the end of validation by inserting or assigning (overwriting) the map
this->parameters_.insert_or_assign(parameter->get_name(), parameter);

// to access the parameter value at runtime, use the map and dynamic cast, rather than a parameter pointer
// instead of this:
this->buffer_size_->get_value();
// do this:
this->get_parameter_value<int>("buffer_size");
```

This avoids the issue of the switch case, and it also reduces the number of individual class properties (which was another intention of the parameter map). However, I do see the preference to use local Parameter pointers so that you don't have to do type casting at runtime. I'm not sure how much it really saves in terms of efficiency but these classes should be optimised where possible.

Therefore I would say keep the switch case here, and in cases where dynamic casting is ok just avoid defining Parameter pointers as class properties.